src/main.rs

```diff
@@ -149,7 +200,16 @@ impl State {
     ///
     /// No change is made to `self` if an error occurs.
     fn do_move(&mut self, mov: Move) -> Result<NextStep, HanoiError> {
-        unimplemented!()
+        let Move{from, to} = mov;
+        let result = match self.peek_disk(from) {
+            // push_disk will not modify state if pushes if valid move
+            Some(disk) => self.push_disk(to, disk),
+            None => Err(HanoiError::EmptyFrom(from)),
+        };
+        match result {
+            Ok(()) => { self.pop_disk(from); Ok(self.next_step()) },
+            Err(e) => Err(e),
+        }
```

Comment:
Hmm... If it's `Ok` do something, else pass the error though... Have we seen this pattern?